### PR TITLE
Fix: no validation for shared policies when closing account

### PR DESCRIPTION
**Pull request by rayane-d**
### Details

Optimistically show shared policy error message when closing the account.
### Fixed Issues

$ https://github.com/Expensify/App/issues/27553
PROPOSAL: https://github.com/Expensify/App/issues/27553#issuecomment-1722450663

### Tests

1. Create a workspace
2. Invite a member to the workspace
3. Go to Settings > Security > Close Account
4. Verify that an error is displayed that say: "You still own workspaces that are shared with other users. Please delete the workspace or have another user take ownership before you try again." in English, or "Sigues siendo propietario de espacios de trabajo compartidos con otros usuarios. Elimina el espacio de trabajo o asigna otro propietario antes de volver a intentarlo." in Spanish.
- [x] Verify that no errors appear in the JS console

### Offline tests

Same as above tests.
### QA Steps

Same as above tests.
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
        - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/77965000/36dbac28-9a98-4d41-9e1f-16b6afb834bd

</details>

<details>
<summary>Mobile Web - Chrome</summary>

</details>

<details>
<summary>Mobile Web - Safari</summary>

</details>

<details>
<summary>Desktop</summary>

</details>

<details>
<summary>iOS</summary>

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

**Comment by sobitneupane** [comment]
@rayane-djouah Sorry for the delay. I was OOO. Can you please resolve the merge conflict. Let's try to get it merged before next week.

**Comment by sobitneupane** [comment]
@rayane-djouah Have you updated the PR as per the [suggestion](https://github.com/Expensify/App/issues/27553#issuecomment-1756727828) in the issue? The screen recordings in the PR doesn't reflect it.

We would link to show error as shown in the screenshot below.

**Comment by sobitneupane** [comment]
Sorry for the delay @rayane-djouah. Can you please raise the discussion in issue. In the meantime, I will see what can be done.

File: src/libs/PolicyUtils.ts [hunk 1]
```diff
@@ -197,6 +197,26 @@ function isPendingDeletePolicy(policy: OnyxEntry<Policy>): boolean {
...
+/**
+ * Checks if the current user has a shared policy.
+ */
+function hasSharedPolicies(policies: OnyxCollection<Policy>, policyMembersCollection: OnyxCollection<PolicyMembers>): boolean {
+    let found = false;
+    if (policies && policyMembersCollection) {
+        Object.keys(policies).forEach((policyKey) => {
+            const policy = policies[policyKey];
+            if (policy && !isPolicyAdmin(policy)) {
+                return;
```
Comment (sobitneupane) [inline src/libs/PolicyUtils.ts:209]:
```suggestion
                return found;
```

File: src/pages/settings/Security/CloseAccountPage.js [hunk 2]
```diff
@@ -82,6 +93,15 @@ function CloseAccountPage(props) {
         return errors;
     };
 
+    const getErrorMessage = () => {
+        let error = '';
+        const hasSharedPolicies = PolicyUtils.hasSharedPolicies(props.policies, props.allPolicyMembers);
+        if (hasSharedPolicies) {
+            error = props.translate('closeAccountPage.hasSharedPolicies');
+        }
+        return error;
```
Comment (sobitneupane) [inline src/pages/settings/Security/CloseAccountPage.js:102]: Should we early return here?
```suggestion
        if (!hasSharedPolicies) {
            return '';
        }
        return props.translate('closeAccountPage.hasSharedPolicies');;
```

File: src/libs/PolicyUtils.ts [hunk 3]
```diff
@@ -197,6 +197,26 @@ function isPendingDeletePolicy(policy: OnyxEntry<Policy>): boolean {
     return policy?.pendingAction === CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE;
 }
 
+/**
+ * Checks if the current user has a shared policy.
+ */
+function hasSharedPolicies(policies: OnyxCollection<Policy>, policyMembersCollection: OnyxCollection<PolicyMembers>): boolean {
+    let found = false;
```
Comment (sobitneupane) [inline src/libs/PolicyUtils.ts:204]: NAB: I think we should rename it to something more specific like `hasSharedPoliciesFlag`.

File: src/languages/es.ts [hunk 4]
```diff
@@ -775,6 +775,8 @@ export default {
         enterDefaultContact: 'Tu método de contacto predeterminado',
         defaultContact: 'Método de contacto predeterminado:',
         enterYourDefaultContactMethod: 'Por favor, introduce tu método de contacto predeterminado para cerrar tu cuenta.',
+        hasSharedPolicies:
+            'Sigues siendo propietario de espacios de trabajo compartidos con otros usuarios. Elimina el espacio de trabajo o asigna otro propietario antes de volver a intentarlo.',
```
Comment (sobitneupane) [inline src/languages/es.ts:779]: For reference: https://github.com/Expensify/App/issues/27553#issuecomment-1768641981

File: src/components/Form/FormProvider.js [hunk 5]
```diff
@@ -172,11 +177,15 @@ const FormProvider = forwardRef(
 
                 if (!_.isEqual(errors, touchedInputErrors)) {
                     setErrors(touchedInputErrors);
+                    // Clear the client side form validation error if any
+                    if (errorMessage) {
+                        setErrorMessage('');
+                    }
```
Comment (sobitneupane) [inline src/components/Form/FormProvider.js:183]: @rayane-djouah Can you please help me understand this change?

**Review by sobitneupane — COMMENTED** [review]
Sorry for the delay @rayane-djouah 

I have few suggestions and concerns.

Overall, changes look good and test well.

https://github.com/Expensify/App/assets/25876548/5853b1a0-ab90-4089-a996-425a6c880cbd

File: src/libs/PolicyUtils.ts (hunk 1 above)
Comment (sobitneupane) [reply src/libs/PolicyUtils.ts:209]: My bad!

File: src/libs/PolicyUtils.ts (hunk 3 above)
Comment (sobitneupane) [reply src/libs/PolicyUtils.ts:205]: @rayane-djouah Sorry for the confusion. I was taking about the `found` variable not the function name. `hasSharedPolicies` is fine for function name.

**Review by sobitneupane — COMMENTED** [review]
### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/25876548/5853b1a0-ab90-4089-a996-425a6c880cbd
</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/25876548/ac76ad59-5319-4601-9dd8-117dc076d8a2


</details>

**Comment by sobitneupane** [comment]
@rayane-djouah Please make sure to include Screen Recording for all platform.

I am facing following error in IOS/safari.

https://github.com/Expensify/App/assets/25876548/6bee6b44-1adc-4b22-bd28-6fe3347a51ae

**Comment by Gonals** [comment]
> @Gonals could you please help on [this issue](https://github.com/Expensify/App/pull/29043#issuecomment-1887867155)? it seems that it originate from the backend.

Hmm. Interesting. Does this happen consistently? I take it is not for all flows?

**Comment by Gonals** [comment]
> > Hmm. Interesting. Does this happen consistently? I take it is not for all flows?
> 
> It is not appearing in the Onyx database. I tried logging it in the dev console at various points in the app, but I am getting an empty object. @Gonals

So, currently, is it never possible to close an account?

**Pull request closed without merging**